[API] Enforce authentication on POST requests for lobbies and chat messages

chrislowzhengxi opened the pull request:
This PR ensures that only authenticated users can create lobbies, post chat messages, and access chat feeds. The following changes were made:

- LobbyListView updated to use IsAuthenticatedOrReadOnly and auto-populate created_by.
- MessageView and MessageFeedView restricted to authenticated users only.
- Added members=[self.user.id] in tests to satisfy model requirements when creating a lobby.
- Confirmed serializers support created_by as read_only.
- Added tests in AccessControlTests to verify that unauthenticated users receive 403 errors and authenticated users can successfully create or retrieve content.


Closes #711 

Test coverage:
All 7 authentication-related tests passed:

```
test_authenticated_user_can_post_lobby
test_unauthenticated_user_cannot_post_lobby
test_unauthenticated_user_cannot_post_game
test_authenticated_user_can_post_message
test_unauthenticated_user_cannot_post_message
test_authenticated_user_can_fetch_message_feed
test_unauthenticated_user_cannot_fetch_message_feed
```

yizou275 committed:
[knowledge-base] Add Backend Moderator Pending Guide List Page (#779)

* [knowledge-base] moderator guide list

Add the page that displays all pending list. And restrict only
moderators to access this url.

For: #749

* [knowledge-base] backend moderator listview

- change ModeratorGuidesPending to generic view
- add sorting

For: #749

* [knowledge-base] fix merge error

* [knowledge-base] fix pre-commit error

---------

Co-authored-by: Yi Zou <yizou@Yis-MacBook-Air.local>
Co-authored-by: David Makoyo <makoyo.david@me.com>

Commit:
[api] Fix spam check crash on None content

Fixes:#709

Commit:
[api] Fix spam check crash on None content

Fixes:#709

Commit:
[api] Modified MessageView in views.py so it passes pytests

Closes: #823

Commit:
[api] Modified MessageView in views.py so it passes pytests

Closes: #823

Commit:
[api] Modified MessageView in views.py so it passes pytests

Closes: #823

chrislowzhengxi commented:
Just added some changes to the tests `test_feed_message`, `test_create_message`, and `test_delete_message` so all the authentication tests passed.

abrodriguezalman reviewed (approved):
LGTM! I have run all tests under the new AccessControlTests and SpamFilterTests on my local machine by running the following commands: "python manage.py test chigame.api.tests.test_api.AccessControlTests" and "python manage.py test chigame.api.tests.test_api.SpamFilterTests". Both commands ran successfully with no failures. The tests themselves seem to be accurately checking for the correct information. I can also confirm that changes to the various API views to add authentication are well implemented.

MuyanXie commented:
Incredible! Do we have anything other than the tests to verify that the spam is working?

chrislowzhengxi commented:
> Incredible! Do we have anything other than the tests to verify that the spam is working?

Sorry for the long comments - this probably could have been more effective with Postman. I feel like things get complicated when I have to manually test them. 

- For the ones that checks regular (i.e. "can") access, please **log in** and go to Go to http://127.0.0.1:8000.  
- For the ones that check unauthenticated users, please run from the root terminal, (with python3 manage.py runserver turned on). 

1. test_authenticated_user_can_post_lobby

Go to http://127.0.0.1:8000/api/lobbies/
```
{
  "game": 1,
  "name": "New Lobby",
  "min_players": 2,
  "max_players": 6,
  "members": [1,2,3]
}
```


2. test_unauthenticated_user_cannot_post_lobby
On terminal (with python3 manage.py runserver turned on) 

```
curl -X POST http://127.0.0.1:8000/api/lobbies/ \
  -H "Content-Type: application/json" \
  -d '{
    "game": 1,
    "name": "Fail Lobby",
    "min_players": 2,
    "max_players": 6
  }

```
What to expect: 
`{"detail":"Authentication credentials were not provided."}`


3. test_unauthenticated_user_cannot_post_game
```
curl -X POST http://127.0.0.1:8000/api/games/ \
  -H "Content-Type: application/json" \
  -d '{"name": "Uno", "max_players": 4}'
```

What to expect: 
`{"detail":"Authentication credentials were not provided."}
`



4. test_authenticated_user_can_post_message
http://127.0.0.1:8000/api/tournaments/chat/

Make sure a tournament has been created already (I did so in the admin site, with tournament id = 1).
And make sure to create a chat (I did this in the shell): 
```
>>> from chigame.games.models import Tournament, Chat
>>> 
>>> t = Tournament.objects.get(pk=1)
Chat.objects.create(tournament=t)>>> Chat.objects.create(tournament=t)
<Chat: Chat for tournament Tournament Testing Content Review: Post Test from 05/23/2025 to 05/29/2025>
```


```
{
  "update_on": null,
  "content": "Hello from an authenticated user",
  "sender": "chrislowzhengxi@example.com",
  "tournament": 1
}

```


5. test_unauthenticated_user_cannot_post_message
```
curl -X POST http://127.0.0.1:8000/api/tournaments/chat/ \
  -H "Content-Type: application/json" \
  -d '{"update_on": null, "content": "This should fail", "sender": "chrislowzhengxi@example.com", "tournament": 1}'
```

Expect: `{“detail":"Authentication credentials were not provided."}`




6. test_authenticated_user_can_fetch_message_feed
http://127.0.0.1:8000/api/tournaments/chat/feed/
```
{
  "token_id": 0,
  "tournament": 1
}

```


7. test_unauthenticated_user_cannot_fetch_message_feed

```
curl -X POST http://127.0.0.1:8000/api/tournaments/chat/feed/ \
  -H "Content-Type: application/json" \
  -d '{
    "token_id": 0,
    "tournament": 1
  }'

```
`{"detail": "Authentication credentials were not provided."}`

MuyanXie reviewed (changes requested):
@chrislowzhengxi The current version looks good to me. Please review changes on the chat/messages feature as it's now deprecated. The team live-chat will now manage the chat feature. Also, please fix the CI test error after merging dev into the current branch

Make sure you are git pull first before making changes

Commit:
Merge branch 'dev' of github.com:uchicago-cs/chigame into api/adding-authentication

Commit:
[api] Fixed 401 errors for unauthenticated users for tests. Closes: #711

chrislowzhengxi commented:
> @chrislowzhengxi The current version looks good to me. Please review changes on the chat/messages feature as it's now deprecated. The team live-chat will now manage the chat feature. Also, please fix the CI test error after merging dev into the current branch
> 
> Make sure you are git pull first before making changes

I’ve merged dev, fixed the CI test failures, and confirmed that the CI tests all pass now (I fixed the 403 error to return 401 instead). Thanks for the heads-up on the chat/messages deprecation — I won’t make further changes to those parts unless requested.

MuyanXie commented:
@chrislowzhengxi Sounds good. Approved!

MuyanXie reviewed (approved):
LGTM!

MuyanXie commented:
This PR is reverted so it will not be graded
